app/controllers/experiments_controller.rb

```diff
@@ -410,50 +419,31 @@ def create
     # Params validation
     type = params[:type] || 'experiment'
     Utils::raise_error_unless_has_key(CONSTRUCTORS, type, "Not a correct experiment type: #{type}")
-    workers_scaling_params = nil
+
     workers_scaling_enabled = (params[:workers_scaling] == 'true')
+    workers_scaling_initializer = nil
     if workers_scaling_enabled
-      message_prefix = 'Missing workers scaling parameter'
-      Utils::raise_error_unless_has_key(params, :workers_scaling_params, message_prefix.pluralize)
-      workers_scaling_params = Utils::parse_json_if_string(params[:workers_scaling_params]).symbolize_keys
-      if workers_scaling_params[:plgrid_default]
-        if InfrastructureFacadeFactory.get_facade_for(:qsub).get_infrastructure_configurations(current_user.id).blank?
-          raise InfrastructureErrors::NoCredentialsError.new('Missing credentials for PlGrid resources')
-        end
-      else
-        [:name, :allowed_infrastructures, :time_limit].each do |param|
-          Utils::raise_error_unless_has_key(workers_scaling_params, param, "#{message_prefix} #{param}",
-                                            'workers_scaling_params')
-        end
-      end
-      # TODO more precise validation
+      workers_scaling_initializer = WorkersScaling::Initializer.new(current_user.id, params)
+      workers_scaling_initializer.validate_params
     end
 
     # Create experiment
-    experiment = send(CONSTRUCTORS[type])
+    result = send(CONSTRUCTORS[type])
 
     # Start workers scaling
-    if workers_scaling_enabled
-      planned_finish_time = Time.now + (workers_scaling_params[:time_limit] || 0).minutes
-      if workers_scaling_params[:plgrid_default]
-        WorkersScaling::AlgorithmFactory.plgrid_default(experiment.id.to_s, current_user.id)
-        experiment.plgrid_default = true
-      else
-        allowed_infrastructures = workers_scaling_params[:allowed_infrastructures].map do |record|
-          # TODO Introduce infrastructure id
-          {infrastructure: {name: record['name'].to_sym, params: record['params'].symbolize_keys}, limit: record['limit']}
-        end
-        WorkersScaling::AlgorithmFactory.initial_deployment(
-            experiment.id,
-            current_user.id,
-            allowed_infrastructures,
-            planned_finish_time,
-            workers_scaling_params
-        )
-      end
-      experiment.workers_scaling = true
-      experiment.planned_finish_time = planned_finish_time
-      experiment.save
+    if workers_scaling_enabled and result[:status] == :ok
+      workers_scaling_initializer.start(result[:experiment])
+    end
+
+    if not result.has_key? :html and not result.has_key? :json
+      flash[:error] = 'Unexpected error, please contact administrator'
+      result[:html] = experiments_path
+      result[:json] = {status: :error, message: flash[:error]}
+    end
+
+    respond_to do |format|
+      format.html { redirect_to(result[:html]) } if result.has_key? :html
```

Comment:
> Tworzenie różnych typów eksperymentów zawiera więcej logiki niż samo wywołanie fabryki.

Nie za bardzo rozumiem dlaczego wiekszosci logiki z tych konstruktorow nie mozna przeniesc do fabryki ?
W przypadku create_experiment i create_custom_points_experiment mamy do czynienia z przygotowaniem eksperymentu, stworzeniem progress bara itp. Na pierwszy rzut oka nie widac powodu dla ktorego tej logiki nie daloby sie umiescic we fabryce.